### PR TITLE
VideoBackends: Use proper floating point depth precision.

### PR DESCRIPTION
**Pull request by Armada651**
Still doing some accuracy tests, but it seems promising that this will fix our off-by-one inaccuracies.

### CODE REVIEW

**Comment by comex** [comment]
Can you explain in more detail what this does?

**Comment by Armada651** [comment]
Hardware tests showed that our depth values are off-by-one in hardware: https://gist.github.com/Armada651/bab3e023b5b813e18cb7#comment-1446986

Using a graphics analyzer I could see that we are accurate to the point where we actually store the depth value. According to Fiora this inaccuracy was introduced by the divisor so I changed that after which the D3D backend passes the hardware test. The OGL backend still fails the test though, but now in the other direction with a value of 13884591.

**Comment by FioraAeterna** [comment]
It's a problem of "correct range, correct precision, but not both".

The only way to losslessly store a 24-bit integer into a 32-bit float is to multiply it by a power of 2 before conversion (or no multiply at all). 0xFFFFFF isn't a power of 2, so error will occur.

This unfortunately also means "max depth" (0xFFFFFF) won't actually map to 1.0. I don't know enough about the graphics stack to know if this matters.

**Comment by Armada651** [comment]
Confirmed to fix snap in D3D, this PR currently breaks OGL because it's now off-by-one in the _other_ direction.

**Comment by Armada651** [comment]
@degasus Since the OGL backend reads the depth buffer as integers this is a bit difficult to fix. It may be fixed if we just use `glClipControl()`, but then we'd still need to implement a hack for older version of OpenGL.

Another option is to switch back to the old divisor and add one where needed like we do with the epsilon hack, but I don't really like adding more epsilons to our code.

File: Source/Core/VideoBackends/D3D/PixelShaderCache.cpp [hunk 1]
```diff
@@ -146,7 +146,7 @@ const char depth_matrix_program[] = {
 	" in float4 pos : SV_Position,\n"
 	" in float3 uv0 : TEXCOORD0){\n"
 	"	float4 texcol = Tex0.Sample(samp0,uv0);\n"
-	"	int depth = int(round(texcol.x * float(0xFFFFFF)));\n"
+	"	int depth = int(round(texcol.x * 16777216.0));\n"
```
Comment (degasus) [inline Source/Core/VideoBackends/D3D/PixelShaderCache.cpp]: After multiplying with 16777216.0, you need to make sure that the result is always smaller than 16777216.0. I think round() is also wrong now, floor sounds better to me.

File: Source/Core/VideoBackends/D3D/Render.cpp [hunk 2]
```diff
@@ -419,11 +419,11 @@ u32 Renderer::AccessEFB(EFBAccessType type, u32 x, u32 y, u32 poke_data)
 		if (bpmem.zcontrol.pixel_format == PEControl::RGB565_Z16)
 		{
 			// if Z is in 16 bit format you must return a 16 bit integer
-			ret = ((u32)(val * 0xffff));
+			ret = ((u32)(val * 65536.0f));
```
Comment (degasus) [inline Source/Core/VideoBackends/D3D/Render.cpp]: May val every be 1.0? ret isn't allowed to be 65536.

Comment (Armada651) [inline Source/Core/VideoBackends/D3D/Render.cpp]: As explained above `val` will never be 1.0.

File: Source/Core/VideoBackends/OGL/FramebufferManager.cpp [hunk 3]
```diff
@@ -89,7 +89,7 @@ FramebufferManager::FramebufferManager(int targetWidth, int targetHeight, int ms
 		glTexParameteri(m_textureType, GL_TEXTURE_MAX_LEVEL, 0);
 		glTexParameteri(m_textureType, GL_TEXTURE_WRAP_S, GL_CLAMP_TO_EDGE);
 		glTexParameteri(m_textureType, GL_TEXTURE_WRAP_T, GL_CLAMP_TO_EDGE);
-		glTexImage3D(m_textureType, 0, GL_DEPTH_COMPONENT24, m_targetWidth, m_targetHeight, m_EFBLayers, 0, GL_DEPTH_COMPONENT, GL_UNSIGNED_INT, nullptr);
+		glTexImage3D(m_textureType, 0, GL_DEPTH_COMPONENT24, m_targetWidth, m_targetHeight, m_EFBLayers, 0, GL_DEPTH_COMPONENT, GL_FLOAT, nullptr);
```
Comment (degasus) [inline Source/Core/VideoBackends/OGL/FramebufferManager.cpp]: s/GL_DEPTH_COMPONENT24/GL_DEPTH_COMPONENT32F/ The last enum is just for the (not given) data

Comment (Armada651) [inline Source/Core/VideoBackends/OGL/FramebufferManager.cpp]: Already have a commit ready for it.

**Comment by Armada651** [comment]
@degasus Ready for review.

Do you still want to switch the D3D depth buffer from `DXGI_FORMAT_R24G8_TYPELESS` to `DXGI_FORMAT_R32_FLOAT`?

File: Source/Core/VideoCommon/PixelShaderGen.cpp [hunk 4]
```diff
@@ -1224,11 +1224,11 @@ static inline void WritePerPixelDepth(T& out, pixel_shader_uid_data* uid_data, A
 		if (ApiType == API_OPENGL)
 			out.Write("\tscreenpos.y = %i - screenpos.y;\n", EFB_HEIGHT);
 
-		out.Write("\tdepth = float(" I_ZSLOPE".z + " I_ZSLOPE".x * screenpos.x + " I_ZSLOPE".y * screenpos.y) / float(0xFFFFFF);\n");
+		out.Write("\tdepth = float(" I_ZSLOPE".z + " I_ZSLOPE".x * screenpos.x + " I_ZSLOPE".y * screenpos.y) / 16777216.0;\n");
```
Comment (degasus) [inline Source/Core/VideoCommon/PixelShaderGen.cpp]: Maybe a bug in the old code, but what shall happen here on ovrflow: clamping or integer overflow? You've changed the behavior from clamping (at the output) to integer overflow (in the consumer functions)

**Comment by degasus** [comment]
@Armada651 If there are no issues, we should use the same format on both backends.

File: Source/Core/VideoCommon/PixelShaderGen.cpp (hunk 4 above)
Comment (Armada651) [inline Source/Core/VideoCommon/PixelShaderGen.cpp]: Won't the GPU already clamp the output to 1.0 since we commit the depth here?

Comment (degasus) [inline Source/Core/VideoCommon/PixelShaderGen.cpp]: Not when we use a float z buffer. Also 1.0 should be treated as a bug, so we need a smaller clamping.

Comment (Armada651) [inline Source/Core/VideoCommon/PixelShaderGen.cpp]: So how about we clamp to 0, 0xFFFFFF before we do the division?

**Comment by ZephyrSurfer** [comment]
The sky is black in the simpsons-tev above.

**Comment by Armada651** [comment]
@ZephyrSurfer We're aware of the regression, but we're probably going to merge it without fixing that bug. Since the sky in simpsons-tev is also black on the Software backend and that game has never worked correctly in Dolphin anyway we don't see a need to fix it right now.

**Comment by degasus** [comment]
LGTM, but please wait for fifoci ;)

**Comment by degasus** [comment]
@dolphin-emu-bot allow-merge

**Comment by degasus** [comment]
@dolphin-emu-bot allowmerge

**Comment by thegrb93** [comment]
Some good and some bad.
https://fifoci.dolphin-emu.org/compare/277017-276312/
https://fifoci.dolphin-emu.org/compare/286660-281278/

Mini ninjas changes looks different between windows and linux
https://fifoci.dolphin-emu.org/compare/283694-280036/
https://fifoci.dolphin-emu.org/compare/286655-281290/

**Comment by BhaaLseN** [comment]
Too bad Windows/DX builder was stuck, so we didn't get this one earlier. Simpsons however was more or less accepted as broken, since it never worked correctly. Def Jam however is odd.

**Pull request merged**